six/modules/c++/six.sicd/tests/test_read_sicd_with_extra_des.cpp

```diff
@@ -0,0 +1,221 @@
+    secondHdr.setField("DESSHLPG", std::string("1", 125));
+    secondHdr.setField("DESSHLPT", std::string(" ", 20));
+    secondHdr.setField("DESSHLI", std::string(" ", 20));
+    secondHdr.setField("DESSHLIN", std::string(" ", 120));
+    secondHdr.setField("DESSHABS", std::string(" ", 200));
+    des.getSubheader().setSubheaderFields(secondHdr);
+
+    //Wrong length
+    nitf::TRE middleTRE(six::Constants::DES_USER_DEFINED_SUBHEADER_TAG,
+            "XML_DATA_CONTENT_283");
+    middleTRE.setField("DESCRC", "99999");
+    middleTRE.setField("DESSHFT", "XML00000");
+    middleTRE.setField("DESSHDT", "2016-06-01T00:00:00Z");
+    middleTRE.setField("DESSHRP", "1234567890123456789012345678901234567890");
+    middleTRE.setField("DESSHSI", std::string(" ", 60));
+    middleTRE.setField("DESSHSV", "Version 10");
+    middleTRE.setField("DESSHSD", "2016-06-01T00:00:00Z");
+    middleTRE.setField("DESSHTN", std::string(" ", 120));
+
+    nitf::DESegment middleDES = record.newDataExtensionSegment();
+    middleDES.getSubheader().getFilePartType().set("DE");
+    middleDES.getSubheader().getTypeID().set("XML_DATA_CONTENT");
+    middleDES.getSubheader().getVersion().set("01");
+    middleDES.getSubheader().getSecurityClass().set("U");
+
+    nitf::SegmentMemorySource middleSource(segmentData, strlen(segmentData),
+            0, 0, true);
+    mem::SharedPtr<nitf::SegmentWriter> middleSegmentWriter(new nitf::SegmentWriter);
+    middleSegmentWriter->attachSource(middleSource);
+    writer.addAdditionalDES(middleSegmentWriter);
+
+
+    //Wrong length again
+    nitf::TRE shortTRE(six::Constants::DES_USER_DEFINED_SUBHEADER_TAG,
+            "XML_DATA_CONTENT_005");
+    shortTRE.setField("DESCRC", "99999");
+
+    nitf::DESegment shortDES = record.newDataExtensionSegment();
+    shortDES.getSubheader().getFilePartType().set("DE");
+    shortDES.getSubheader().getTypeID().set("XML_DATA_CONTENT");
+    shortDES.getSubheader().getVersion().set("01");
+    shortDES.getSubheader().getSecurityClass().set("U");
+
+    nitf::SegmentMemorySource shortSource(segmentData, strlen(segmentData),
+            0, 0, true);
+    mem::SharedPtr<nitf::SegmentWriter> shortSegmentWriter(new nitf::SegmentWriter);
+    shortSegmentWriter->attachSource(shortSource);
+    writer.addAdditionalDES(shortSegmentWriter);
+
+    std::auto_ptr<TempFile> temp(new TempFile());
+    writer.save(&bandData[0], temp->pathname());
+    return temp;
+}
+}
+
+
+int main(int argc, char** argv)
+{
+    try
+    {
+        validateArguments(argc, argv);
+        const std::string xmlPathname(argv[1]);
+
+        six::XMLControlFactory::getInstance().addCreator(
+            six::DataType::COMPLEX,
+            new six::XMLControlCreatorT<six::sicd::ComplexXMLControl>());
+
+        std::auto_ptr<TempFile> nitf = createNITFFromXML(xmlPathname);
+        six::NITFReadControl reader;
+        reader.load(nitf->pathname());
+        std::cout << "Test passed\n";
```

Comment:
I think this is a good test but one thing is that you're just verifying reading didn't error out... not checking that we actually read the right thing out.  So after you load(), if you can check that the `ComplexData` looks as expected, you should be good.

Note that you don't want to do just a string comparison of the XML because Linux and Windows will format things differently unless you have a different XML file for each platform.  But you should be able to use some of those `six::sicd::Utilities` to quickly go from XML string to `ComplexData` and then use the equality operator to compare.